### ISSUE TITLE
Ability to Associate issues with a Version

### ISSUE THREAD

**Issue by dazinator** [CatenaLogic/JiraCli#4]
Hi there! I was looking to use this to Create a version for each Team City CI build. However' i'd also like to associate a set of Jira issues with that particular version. I am able to get a list of issues based on check-in comments / git log, but the bit I am stuck at is how to then assiciate those Jira issues with the Version that is created. May I ask how you achieve / manage this currently - is it a manual process to tie up the issues in jira with the version / release?

**Comment by GeertvanHorrik** [comment]
We actually use this for example the same task.

1. Use GitVerison to generate the version
2. Create just released version if it's not there yet so users can report issues on this version
3. When an official version has been released, merge all sub releases into the big one and release it in JIRA as well

What steps are giving you issues? I can easily give you all the commands that we use.

**Comment by dazinator** [comment]
Thanks - yep this exactly what I am going for so this tool is perfect.

The only bit I am struggling with, is that Jira issues in our sprint start off not assigned to any version. When the developer progresses the issue, they make a number of commits (including issue number in commit log). This triggers a new CI build (unstable) - and this will end up creating a new Version in Jira for the CI build (thanks to this tool). However what associates the issue the developer was working on to the created CI version in Jira? I was thinking about automating that based on check in comments, but it sounds like that will have to remain a manual process - i.e a developer commits he work for the issue. waits for the CI build to complete, and then goes into Jira and assigns the issue to the just created CI version?

Great to hear that this tool also does the merge process when we create a stable release - I was thinking we'd have to do that manually :)

**Comment by GeertvanHorrik** [comment]
We use 2 versioning:
1. FixVersion => always vNext (stable). This means that we will always put a FixVersion (as planning) on a major release
2. AffectsVersion => this contains the version the issue occurred in. This is always a released version (most probably an unstable build automatically created by this tool).

This means we don't necessarily set the "fixVersion" to the version that has been released. However, we do have another system that generates release notes for us. It uses some "assumptions" to use the 'resolved issue date' and the creation dates of the versions. If an issue was resolved between 2 releases (automatically created by JiraCli), we assume that the issue was fixed in the last build and thus show it as release notes in our unstable versions.

As soon as we release a stable version, we can just show all issues added to the major version milestone since this tool will merge all versions (so 1.0.0-unstable1, 1.0.0-unstable2 will all be merged into 1.0.0).

**Comment by dazinator** [comment]
Ah ok I understand. So basically for vNext work you don't associate issues with particular CI versions. This is what I'd like to achieve really. You can use a seperate tool with some clever date based logic to link resolved issues to an appropriate CI version for the purpose of release notes. I'd like to avoid this and do it all through Jira.

When we then release, we will do the same as you in that we will merge the CI versions into the stable version. 

Could you see the ability to associate issues to versions being added to this tool? I'll probably end up trying to do this in my Fork at some point, and will happily submit a PR if desired, but otherwise I can just leave it in my Fork ;)

What i'd like to do is add in a command for associating a list of issues numbers to a version number.

In my Team City CI build process, i'll look at issues (based on check in comments) associated with the build, and i'll get a list of those issue numbers. I'll then call this tool, after creating the version, and associate those issues with that version..

**Comment by GeertvanHorrik** [comment]
Happy to accept such a PR, no need to maintain a custom fork :-)

This tool can actually already associate issues to a version (we do that when merging versions), so you could reuse that part to implement your action / feature.

**Comment by dazinator** [comment]
I did look at the merge functionality for assigning issues, but I couldn't re-use it in the end, as it looks like it actually does this by deleting the old versions, and as part of that you can tell Jira to move the associated issues to a new version:

```
public static void DeleteProjectVersion(this IJiraRestClient jiraRestClient, JiraProjectVersion projectVersion, JiraProjectVersion projectToMoveFixIssuesTo = null,
            JiraProjectVersion projectToMoveAffectedIssuesTo = null)
        {
            Argument.IsNotNull(() => jiraRestClient);
            Argument.IsNotNull(() => projectVersion);

            var resource = string.Format("rest/api/2/version/{0}", projectVersion.Id);

            if (projectToMoveFixIssuesTo != null)
            {
                resource += resource.Contains("?") ? "&" : "?";
                resource += string.Format("moveFixIssuesTo={0}", projectToMoveFixIssuesTo.Id);
            }

            if (projectToMoveAffectedIssuesTo != null)
            {
                resource += resource.Contains("?") ? "&" : "?";
                resource += string.Format("moveAffectedIssuesTo={0}", projectToMoveAffectedIssuesTo.Id);
            }

            var responseJson = jiraRestClient.ExecuteRequest(Method.DELETE, resource);
        }

```

In my case, I don't want to replace / delete any version as I'm not merging. So in the end I had to use the update Issue API, as you will see in the PR. This wasn't too bad, was a bit fiddely getting the json syntaxt correct.

**Comment by GeertvanHorrik** [comment]
I think this one can be closed?

**Comment by dazinator** [comment]
Yep - closed via [PR#7](https://github.com/CatenaLogic/JiraCli/pull/7)
